### ISSUE TITLE
Collapsable headers please! [feature request]

### ISSUE THREAD

**Issue by RFImages** [giuspen/cherrytree#2389]
Hi I had previously made a request for collapsable text sections in the body of a node or subnode, similar to the way you collapse nodes and subnodes in the node/subnode pane. 

Recently I discovered by accident that Microsoft Word already has this feature so I am using it to edit LONG subnode bodies, then having to copy it back to Cherry tree:
 https://support.microsoft.com/en-us/office/collapse-or-expand-parts-of-a-document-701786e0-95e2-40bf-bfe5-f0233cd3520c

The way they do it is perfect, just identifiy heading lines (h1, h2, h3 as you already do), then when hovering over the heading, an arrow appears on the left that can be clicked to collapse or expand the body below tcontained within that heading. 

This is SOO helpful for super long text nodes, to be able to focus on the area of interest without all the clutter of the rest of the text body. It is so useful that I am willing to donate a downpayment of $50 if you also think it is a feature worth implementing, and then larger donation on completion, depending on how well done and how soon.
What do you say Giuspen?  I needed this yesterday as I strugggle daily with a particularly long To Do list in my todo node and I don't want to have multiple subnodes just to maintain a heirarchy (too many clicks, too time consuming

**Comment by RFImages** [comment]
ok to get attention I will up my bid to $75, payable immediately upon acknowlegement of this feature request and an estimation of when it will be available. When implmented I will donate significantly more!
Yes, I am bribing you! The utility of this feature is emormous for me, and I assume others, since Microsoft saw fit to incorporate it into Word.

**Comment by RFImages** [comment]
:(

**Comment by giuspen** [comment]
Hi @RFImages don't be sad it's difficult for me to follow all the issues I'm only one :)
In fact I apologise to who doesn't receive an answer and feels ignored, I hate that but believe me I become crazy if I follow everything and I fail to focus on a development feature.
I will have a look after v1.0.3 but before this I need to implement virtual nodes.

**Comment by RFImages** [comment]
:) Ok cool, I understand you are packed with work on this great product. As promised I will now donate $75 on good faith.
What are virtual nodes?

**Comment by giuspen** [comment]
Thank you my friend :)
Virtual nodes are a group of node "views" that point to only one node content, so you can see and edit from all the nodes scattered in the tree that belong to that group.
Not sure if there's a better way to call them, I'm still thinking and open to suggestions.

**Comment by RFImages** [comment]
It may not be $75 considering various sources for exchange rates, but will make it up to you generously when this feature is incorporated. Thanks for considering it. Others have asked for it as well. Keep me informed!

**Comment by giuspen** [comment]
It's discussed here https://github.com/giuspen/cherrytree/issues/1222 and has been requested time ago privately from one of the few supporter that have a recurring monthly donations, I cannot make him wait any longer

**Comment by RFImages** [comment]
Oh no I understand I am not asking that, I am just curious about what virtual nodes are. Thanks for the link.

**Comment by RFImages** [comment]
Ok I think understand it now and yes I agree that would be useful. I will comment in the other post.

**Comment by RFImages** [comment]
This got off topic, but wondering where the collapsible header feature might be in the queue?
I am noting other similar requests here
Collapse, Expand Text Selection #2026
[Collapse, Expand Text Selection · Issue #2026 · giuspen/cherrytree (github.com)] (https://github.com/giuspen/cherrytree/issues/2026)_
Collapsable Text Selections #2360

[Collapsable Text Selections · Issue #2360 · giuspen/cherrytree (github.com)](https://github.com/giuspen/cherrytree/issues/2360)

**Comment by giuspen** [comment]
Hi @RFImages I didn't forget about this, I am still to release a bugfix version v1.1.2 today or tomorrow related to the search, then I will be off until half of April without a computer, only phone, and then I will start looking into this.

**Comment by RFImages** [comment]
Thank you @giuspen You mean you can't develop software on a phone? haha
Come to think of it, being able to view cherrytree on android would be great! How will you view your personal cherrytree?
Safe travels to the desert island!

**Comment by giuspen** [comment]
Hey @RFImages I haven't found yet a way to hide part of the GTK text view in order to achieve the result using the gtk3 library interface.
One quick but not as good possibility would be to allow the so called code boxes to collapse. A code box though can currently contain only plain text (this may change in future where I will try to support rich text too)
So basically you would place your header and below a code box with plain text that can collapse and will remember the collapsed status.
Let me know if this may help your workflow or not.

**Comment by giuspen** [comment]
Interesting just after writing to you I found https://github.com/aeischeid/gedit-folding it looks like it is possible to hide sections of text using text tags, I'll investigate further

**Comment by RFImages** [comment]
I am so happy to see this get attention! Especially since I just started a new job and it will be a hug help in keeping me organized and efficient. Thanks for remembering. I just sent a donation 25 to keep you motivated. Paypal would not let me make it monthly for some reason but I will make this recurring.
The code box solution does not help much because all of my text body is highly formatted. I hope that the  text tags work!

**Comment by giuspen** [comment]
Many thanks for your donation @RFImages !
I will complete first the auto/guided recovery of corrupted documents from the most recent backups and then I am on this.
I now know that the library supports what I need and I have a plan so I don't expect problems.
I will update this ticket soon with likely a version for you to try.

**Comment by RFImages** [comment]
Beautiful!
Do you think it will work with the headers as the collapse/expand point similar to the way Word handles it? 
Also I am always afraid to upgrade for fear of my version 1.0.0 .ctz file being corrupted by the newer version. Is my fear unfounded?

**Comment by giuspen** [comment]
Yes the idea is to improve the current parser that adds an anchor before every header when generating the TOC so that these anchors will be clickable and capable of hiding/showing the text under the header (up to the next header of the same level).
Theoretically the current stable version v1.1.3 is the safest ever as I'm constantly trying to make it more and more robust, but it is certainly good to save a time based backup from time to time. I have always done it even though have never needed one.

**Comment by giuspen** [comment]
In case of a collapsed header, if you close and reopen cherrytree would you expect the header to be still collapsed or expanded?

**Comment by RFImages** [comment]
I think I would prefer it were left in the state it last was when saved (collapsed)

**Comment by RFImages** [comment]
I just had a thought I wonder how this woud affect the search function. If the word is found within a header would it have to expand that header?

**Comment by RFImages** [comment]
Hello @giuspen It has been a couple of months and I am wondering, 
1) Are you ok?
if (you are ok) then
       Is this still in progress? ;
       elseif (you have had time)
                   {Are there obstacles? ;
                     Is it very time consuming and not enough time has elapsed?}
             elseif (not done)
                     I understand, I just hope this doesn't get pushed too far on the back burner. I have need of it every day! ;
             else  I am ready to make monthly contributions! ;
else 
      Thoughts and prayers!

I thought I would ask this in language you understand ;) Feel free to correct my psuedo code syntax.

**Comment by giuspen** [comment]
Hey @RFImages no problema just some much needed holidays. I have started looking at this and will propose a first version soon. Likely the first version will not remember collapsed/expanded status.

**Comment by RFImages** [comment]
Great glad you got some time to refresh! I would be happy to test a handicapped first version.

**Comment by DiagonalArg** [comment]
I'd also be interested in this.  Having used org-mode in emacs, I'd suggest to consider what happens when we do a search.  In org-mode, when we do a progressive search, the subsection opens at the identified string/regexp, and then closes when we move to the next in the progressive search.  It only stays open if we begin editing there.  

So, perhaps you'd want to open the drop-down when we click on one find in the search box, and then close that/ open the next when we click on that.

I hope I've explained that clearly enough.  

And @giuspen / you've been quite the trooper all these years.  It's been a while since I've checked in here, but you were always remarkably responsive.  Back when you were working on markdown/ pandoc/ lua etc., you have someone working with you.  I'm sorry to hear you're on your own now.  :heart:

**Comment by RFImages** [comment]
Agree, @giuspen is an example of what any software developer should be. As a solo act his support far exceeds that of very large software companies who don't care a whit about the user. With this feature I intend to support him with a regular donation. I hope other users will choose to do the same.

**Comment by giuspen** [comment]
I'm almost there, I will be able to propose a first version as a matter of days.
The direction I'm taking is to remember the expanded/collapsed status between sessions.
About the search that auto expands and then collapses again, it makes sense but please be patient that may come afterwards and not straight away.

**Comment by RFImages** [comment]
Thank you! I sent another donation today.

**Comment by giuspen** [comment]
Thanks @RFImages that's very generous of you!

**Comment by giuspen** [comment]
First version (1.1.4+23) is available to try at https://www.giuspen.net/cherrytree/#testing
I have uploaded the windows version while I'm still preparing the appimage, will be available in a moment
Please test it and provide feedback

**Comment by RFImages** [comment]
I am not sure  how to test this, I downloaded the cherrytree_1.1.4.0+25_win64_portable.7z archive and installed and the "about" says 1.1.4  (nothing about +25) as it did before installing this.

How are collapsible headers exercised? How do you generate the TOC?

**Comment by RFImages** [comment]
Hmm I am only seeing a trailing anchor , and nothing with 1.1.4 (nothing about +25)  in the about:1.1.4 (nothing about +25) 
<img width="677" alt="Screenshot 2024-10-03 104926" src="https://github.com/user-attachments/assets/0d5d4aba-d34c-4b5f-947e-f974bbdc4382">

**Comment by giuspen** [comment]
Hey @RFImages I'm running a bit late so I will release v1.2.0 shortly and you can give me feedback trying the official version

**Comment by RFImages** [comment]
My prelininary feedback is that inserting a global TOC appears to need to place an anchor on every line throughout the file, which I don't think is useful and adds clutter.  Is there a way to at least limit that to Header lines?

**Comment by giuspen** [comment]
The anchors are added only on the line of the headers.
![image](https://github.com/user-attachments/assets/d98c19d6-4bd2-4673-b91a-ea4f4223615e)

**Comment by RFImages** [comment]
Well that looks very great!  What I meant was the anchors you get when you "insert anchor". Maybe I am using TOC wrong, I never use it but when I did it placed anchor at the end of every line in my file.
Perhaps because I selected "All the tree " when I did Insert TOC?

**Comment by giuspen** [comment]
That is weird, I suspect every single line must be marked as a header H1,H2,..,H6
Try to clear formatting for the normal text between headers or send me privately some of your data / test data for me to reproduce

**Comment by RFImages** [comment]
I'm a bit confused, because it is not EVERY line as I originally thought, but a lot of them. I am not sure what the parameters are for inserting anchors by the TOC, but I think it best to wait for the official v1.2.0 release to test this. I don't want to derail you until I know what I am talking about. Thanks for adding this feature I can't wait! It will be infinitely useful.

**Comment by giuspen** [comment]
v1.2.0 is out now, I still have to build some binaries/packages but windows files are available already.
Currently every header H1 to H6 becomes collapsible and reachable by TOC hyperlink.

**Comment by RFImages** [comment]
Wow this is a life changer for me! I love it, you implemented it just as I imagined. I love that it collapses and expands heirarchically. THANK YOU!
Let's start with I made another 20 euro donation (my initials are R.F.)and will do that monthly, but was a bit uncomfortable doing it automatically through Paypal. I will bump that up as special features are added.
This will really unclutter my lengthy nodes.
I did find a few bugs, I will soon post instructions on how to replicate, 
Off hand, intuitively I would prefer the "eye open" symbol to mean "header open" (the reverse of what you have), but that is a nit.

**Comment by RFImages** [comment]
A warning, some of the bugs I have found will cause data loss. This is not yet ready to test on valuable nodes, unless duplicated or backed up first.

**Comment by giuspen** [comment]
Thank you for your generosity @RFImages 
Certainly if you help me replicate the issues I will sort those out. About the eye open/closed I simply want to follow the standard, maybe comparison with other tools would help.

**Comment by giuspen** [comment]
About the possible data loss, the way data is hidden is via tags/ formatting, so clearing the formatting also removes the invisibility tags

**Comment by RFImages** [comment]
It may not be loss but it did rearrange a few things. Still experimenting.
By invisibity tag do you mean the eyeball "anchor" at the end of each header line? I was calling that a collapser.
That was one weird thing that I noted, - "remove formatting" does change headers back to normal text, but it does not remove the eyeball collapers. Preferably it should.

**Comment by giuspen** [comment]
The collapser is just a clickable placeholder, the tag of invisibility is added/removed to the text following the collapser until the subsequent header of the same type (e.g. H1 to H1 or H2 to H2).
Please keep testing and write down a list of the issues you find.

**Comment by giuspen** [comment]
The rich text tags produce a visible effect like bold, italic, foreground colour etc.
In this case, this particular tag has the effect to hide the text that has the tag, and this is to achieve the effect of the header collapsed / hidden.
Note that if you search for text that is under the invisible tag, the invisibility tag is removed to show you the search result.
Note that if you hover over one of these eye open/closed, the tag name contains the type of header (e.g. h1-number h2-number, ... , h6-number)
the tag is applied from the beginning of the line following the header to the last line before the following header of the same type or the end of the text in the node

**Comment by RFImages** [comment]
I see, thanks. 
Questions/comments:
How often would the TOC need to be re-run? What if a new header is added for example, or the header state is changed say from H1 to H2. What is the procedure for updating the structure (eyeball placement, association with a header line? I would expect remove all formatting then run TOC again. But remove formatting does not remove the eyeball.

This seems a bit dangerous, becasue it is possible to edit the eyeball as though it were a character (delete it or move it - copy paste for example) which destroys the structure. If that happens how can it be cleaned and updated? Again I would expect remove all formatting then run TOC again. But remove formatting does not remove the eyeball.

As it is, there seems to be no way to clean it up and start over, which also makes debugging difficult, as it is possible that the eyeball becomes dissociated from the header that it was originally associated with, or removed entirely.

Suggestion would be to make it impossible to edit the eyeball or break its link to the header, cause the eyeball to track any change in header state of the line. For example, if the header formatting is removed, then also remove the eyeball. This will make it less confusing and disallowing getting into a weird state where eyeballs appear on non-header lines.
Remove formatting should also remove eyeballs.

**Comment by giuspen** [comment]
You do not need to remove formatting in order to re-run the generation of the TOC.
The algorithm should update all the existing anchors (eyeballs) and add those missing.

If by mistake you remove an eyeball you can either Ctrl+Z (undo) or simply re-run the TOC generation.

Maybe it would be useful to add a command that clears the TOC. This will remove all the eyeballs and expand all the collapsed sections.

Ideally yes I should add more code to automatically update the eyeballs after the headers but that is non trivial, will take time, I will improve this gradually, cannot dedicate all the time exclusively on this as there are many other important tasks waiting

Sure I will invert the open/close as you prefer, if somebody else will be not happy I'll make it configurable

You are right it is not very robust at the moment, you could cut and paste an eyeball anchor away and it would still maintain the old property until it is re-generated. It is something I will have to improve

**Comment by RFImages** [comment]
Oh I understand there are other priorities and I greatly appreciate your time with this. Certainly it can be improved (feature creep!) but I just want to be sure it is bug free and usable while your attention is on it. I think now that I understand how it works, perhaps there are no bugs. It has gotten into some weird states but likely from my own misunderstanding. I will continue to work with it to see if anything serious crops up (such as data corruption). Thank you so much for this!  I will continue to comment on the thread but I understand other priorities, please do not feel pressured.

**Comment by RFImages** [comment]
@giuspen I have a question to aid in my testing of this. I am getting into more and more complex nodes that I have accumulated over the years and trying to organize them with collapsible headers.

One big problem I have encountered is nodes that contain URL links. 
Apparently when I copied links and conversations from a Teams chat, it copied the text with a header property set for some reason. So it thinks the URL link is a header and adds a collapser. When I try to clean it with "remove formatting" it also removes any URL link properties and changes it to plain text. I don't think it should do this. Example:

![image](https://github.com/user-attachments/assets/0a0b893f-b07b-4e26-92e1-15886a171e7c)

How can I work around this and simply remove the header property from the URL link? Preferably it will not convert links to plain text, only remove the header property. I don't consider an active URL link to be "formatting" that should be removed. I would like to clean all of these nodes of headers, but that also converts all of my URL links to plain text. This means a ton of manual cleaning, which will take me a lifetime..

**Comment by giuspen** [comment]
Hi @RFImages , a quick way to remove the heading formatting from a line of text is currently to put the cursor at the beginning of the line and press Ctrl+1 either once if it was already exactly a H1, otherwise twice if it was a H2..H6 property, in this case will change to H1 the first time and second time it will remove it.

About the clear formatting not to remove links, it is a valid point, I will have a think and see if I can just do that or at least give the option to the user.

About determining the Heading property hovering (with a tooltip), I'll have a think. I could enable it, I'm just not sure whether most of the users would be annoyed from it or appreciate it. Maybe I'll give that a try.

**Comment by RFImages** [comment]
Thanks for your helpful reply. I had put aside testing becasue ALL of my nodes are infested with links that get marked with eyeballs as if they were headers. I cannot really use the collapsing feature  unless there is a way to cleanse the node of eyeballs for links, or to block them from being put in by the TOC to begin with. I have long nodes with tons of links.

I notice the Ctrl+1 removes the header property but not the eyeballs. There seems to be no quick way to remove or block eyeballs for URL links. Any chance of this being corrected in the near future?

Regarding the tool tips with header information or other properties for the line, I don't think it would be distracting if it takes a while for it to appear and only if the cursor is at the begining of the line. On the other hand it doesn't have to be a tool tip if there were some properties window showing the information.

**Comment by RFImages** [comment]
Hello @giuspen I see there has been a commit associated with this, have you had a chance to look at the undesirable "adding eyeballs to URL links as if they were headers" issue addressed in my previous comment?
I am about to test this with my main data base but wondering if I should hold out until that issue is resolved. Or maybe this commit resolves it? Thank you for your diligence. I will also resume my monthly contribution once this is resolved.

**Comment by giuspen** [comment]
Hello @RFImages I have been busy mainly with another issue I'm trying to deliver ( https://github.com/giuspen/cherrytree/issues/2354 ) so have tried to not get distracted with other issues.
It is not clear to me the situation you have with I assume many links marked as headers that so get picked up at TOC generation time, maybe you can send me a sample of your data so I can understand better what to advise you / how I can help?

**Comment by RFImages** [comment]
Of course I understand the need to focus that is what I ues Cherry Tree for!
I am now actively testing the collapsible headers with version 1.2.0 - I found that the link problem I described above only occurs when URL links are copied from Microsoft Teams chat. I don't know why but not a huge thing.

My biggest fear is data loss and I have encountered a few situations where text is not visible so I am blind to what I am doing and have accidentally deleted things but I have learned what NOT to do to avoid this. I will try to find simple cases to demonstrate. There are some dangerous bugs, so this needs to be used with caution or with knowledge of what NOT to do for now. Not a complaint, just a heads up! But for now this feature  is HUGE for me and I am so happy I just made another donation of $50 euro. CherryTree is an evolving masterpiece! So intuitive and functional. My infinite compliments.

So when you are done with what you need to do, I would greatly appreciate a few very helpful things regarding the collapsable header feature, which will help in efficiency and in my testing:
**In order of importance:**

1. Expand/collapse selected area - minimally collapse all and expand all, but selected area can acheive this by first selecting all 
2. Clean selected area - reverses action of TOC and headers collapsers (removes collapsers)
3. I like the eyeball collapser in theory, but I am finding it difficult to see if it is an open or closed eyeball - especially on a high resolution monitor. The opened and closed eyeballs look very small and similar. I highly suggest using the same down and right arrow symbols that you use in the node tree (down arrow open, right arrow collapsed, or in this case perhaps left arrow would work better since the collapser is a postfix). 
4. The cursor should change from a text select cursor to a pointer while hovering over the collapser. It is hard to target it with the text select cursor.

The focus of my testing is on situations that could cause corruption of the node or data loss.
**Areas of concern:**
Activitys while nodes are collapsed. For this reason, certain actions currently should be avoided while there are any collapsed nodes:

1. Moving text up or down with alt (up/down arrow keys) - this is a big one - when passing text through a collapsed node it causes all kinds of confusion  (blocks of text become invisible) workaround: DO NOT USE TIS FEATURE WHEN A NODE IS COLLAPSED DO THIS! Ideally it should skip over the collapsed node, currently it seems to go line by line through the collapsed contents (even while you can't see what it is doing) and in the process turns on some unwanted invisibility within.
2. Adding/deleting new headers and rerunning insert TOC after it has already been run. So far it seems to be ok to do this.
3. Clicking a hidden subheader in the TOC that is hidden within a collapsed parent header doesn't open the collapsed header to go to the header that was clicked. It should open the parent header and then go to the subheader clicked on the TOC.
5. Searching for the that is contained within a collapsed node. So far this seems to work properly - it opens the collapsed node, then goes to the line that contains the text. (highlighting the text would be even better)

Thank you for your attention to this, I know we would both like to put this feature request to rest (so I don't keep pinging you haha).

**Comment by RFImages** [comment]
Ok sorry to keep pinging you but this is important - I have been actually putting this to use and have found some dangerous situations so I think you should be warned, perhaps this feature is not ready for prime time.
In the process of my normal daily todo list editing, some entire collapsed nodes have vanished. I am quite sure the data is there, but invisible. I know this by turning on line numbers and seeing gaps between the sequencing of line numbers (examples below) This is why an option to "cleanse" the node is important until this can be trusted. This means expanding all nodes, turning off all invisibility, and returning to a state before the TOC in inserted.
Fortunately I have been regularly duplicating the nodes as backup in case something like this occurs. Below is a clip of my current node vs the same section on my bakcup node. Becasue of my editing the line numbers will not be the same, but look at the white spce with gaps in the line sequencing! This tells me that there is hidden data and hidden nodes that are collapsed.  I don't know how or when this occured but a large node is invisible - I have two major H1 headings in red italic underline and simply dissappeared while I was editing elsewhere - look at the line numbers in the white space! :
Before:
<img width="815" alt="Screenshot 2025-01-14 085540" src="https://github.com/user-attachments/assets/d7e702ca-ff3d-45bc-bfcf-b9170679b2ca" />


After:
<img width="811" alt="Screenshot 2025-01-14 085348" src="https://github.com/user-attachments/assets/00096f09-7c8b-4917-884a-84cca7b3059b" />

**Comment by giuspen** [comment]
Hey @RFImages I have not forgotten about this, I will be back following your testing results as soon as I close down https://github.com/giuspen/cherrytree/issues/2354 - one to two weeks time I reckon - I cannot keep bouncing between issues or I will have a breakdown

**Comment by RFImages** [comment]
Of course! I knew that to be the case please finish what you are doing. I am just putting my notes here of what I discover. Overall this feature will be so helpful to me that I must admit I am anxious 😊

**Comment by RFImages** [comment]
Not to pressure, just a note that there is some instability here as well. Twice I have experienced a crash where cherrytree simply dissappears, the most recent was caused by clicking on an eyeball to expand an h3 header. When I reopened cherry tree, I tried the same thing again and the expansion occurred without problem. It appears there is an accumulation of some sort that hits a limit.

**Comment by giuspen** [comment]
Hello @RFImages I had a review of your latest comments and created some notes about your findings.
I will paste here my notes with numbering so I can refer to a specific point.
I will have a look at the icon at first (maybe the size too) and the cursor change when hovering (points 3 and 4).

1)
Expand/collapse selected area - minimally collapse all and expand all for selected area, but selected area can achieve this easily by first selecting all. collapse expand selection would be so much more useful than a global collapse/expand.

A collapsible text section needs a start and an end.
Currently we are using for a start any header (H1 to H6) and for an end another header of the same type or the end of the text in the node.
What you are asking here is to be able to select some text and quickly make it collapsible instead of needing to generate the TOC for the whole node?
Note that in order to achieve the result, I would have to automatically inster a header before and after the text, so the user would still need to select the header type 1 to 6, and the text ti insert into the header


2)
Clean selected area - reverses action of TOC and headers collapsers for the selected area (expands all selected nodes and removes collapsers)

OK


3)
Collapser icon - I like the eyeball collapser icon in theory, but in practice I am finding it difficult to see if it is an open or closed eyeball - especially on a high resolution monitor. The opened and closed eyeballs look very small and similar. It is also difficult to target, especially without a pointer cursor.  I highly suggest using the same down and right arrow symbols that you use in the node tree instead (down arrow open, right arrow collapsed, or in this case perhaps left arrow would work better since the collapser is a postfix). Ideally this could be a preference in the preferences menu!
Also it would be preferable for it to be at the beginning of the line instead of the end because line lengths vary and therefore they are not vertically aligned, but understand it might be more difficult to implement that way, and conflict with other features like bulleting, so this is not huge.

It is tricky to put it at the beginning, I will test the other icons of arrow, maybe it would be useful to have the size configurable?


4)
Cursor change when hovering over collapser - Something odd here - I thought I noticed the cursor change to a hand with index finger while hovering over the collapser but most often it does not change from the text bracket cursor, which makes it very hard to target the collapser. Perhaps the collapser is so small with my high res monitor that I just cannot find the target field. If that is the case the target field needs to be bigger and it would be helpful to assure a cursor change to something pointy (index finger hand or arrow)- the eyeball gets very small at high resolutions. The Expand/Collapse selected area suggested above would be helpful for this problem.

I will check that the cursor changes correctly, I think at the moment it doesn't


5)
Moving selected lines up or down with alt (up/down arrow keys) - this is a big one - when passing selected lines through a collapsed node it causes all kinds of confusion  (blocks of text become invisible) workaround: DO NOT USE THIS FEATURE WHEN A NODE IS COLLAPSED ! Ideally it should skip over the collapsed node, currently it seems to go line by line through the collapsed contents (even while you can't see what it is doing) and in the process turns on some unwanted invisibility within.

It is a valid point. Additional automatic checks must be done before performing this action, maybe expanding automatically the section if currently collapsed


6)
Adding/deleting new headers and rerunning insert TOC after it has already been run. So far it seems to be ok to do this.

This should be working OK


7)
Clicking a hidden subheader in the TOC that is hidden within a collapsed parent header doesn't open the collapsed header to go to the hidden subheader that was clicked. It should open the parent header and then go to the subheader clicked on the TOC.

If a section of text is hidden, there is nothing you can do, you must show it before anything that is inside is working, including a nested collapser


8)
Searching for text that is contained within a collapsed node. So far this seems to work properly - it opens the collapsed node, then goes to the line that contains the text. (highlighting the text would be even better)

I did a test on the current windows version (pre v1.3.0) and it does select properly after expanding, I will need a practical example maybe with test data on this


9)
Copying and pasting a collapsed node works but is not without bugs. Sometimes there is hidden (or deleted?) text after the paste.  I have not yet come up with a test case.

this is tricky, yes a test case would help

**Comment by RFImages** [comment]
Hello @giuspen, thank you for your attention to this, I am sorry I have not been able to give it attention back as I have been moving my office which reuired all the attention I had.

My comments to your comments below:

_I will have a look at the icon at first (maybe the size too) and the cursor change when hovering (points 3 and 4)._

Great I think that is the most annoying thing ATM.
1.
_What you are asking here is to be able to select some text and quickly make it collapsible instead of needing to generate the TOC for the whole node?_

I was not clear. What I was asking for minimally is a right click option to collapse all/expand all feature for the entire node or subnode page, similar to what you have for the node tree. For additional  functionality and usefullness, especially for debugging, expand/collapse only whatever headers are selected. (ignoring any selected lines that are not headers) This would be after running the TOC and the collapers have already been inserted.

i.e. for "collapse selected area"it would only collapse whatever header/collapsers are contained in the selected area, ignoring non-header text that is selected. For "expand selected area" It would expand any header/collapsers that are selected, ignoring any selected non-header text lines. I hope that makes sense.
3.
_It is tricky to put it at the beginning, I will test the other icons of arrow, maybe it would be useful to have the size configurable?_
I thought it might be tricky, that is no big deal. But at least a few spaces at the end before the collpaser appears would make it easier to target.
5.
_It is a valid point. Additional automatic checks must be done before performing this action, maybe expanding automatically the section if currently collapsed_
Yes I think expanding it would be ideal.

I will keep my eye out for test cases for 8. and 9.

Thanks again!

**Comment by RFImages** [comment]
HI @giuspen I have been working quite a bit with this and one thing stands out that makes it confusing or non-intuitive.

You mention above: 
_A collapsible text section needs a start and an end.
Currently we are using for a start any header (H1 to H6) and for an end another header of the same type or the end of the text in the node._

I beleive this should be changed to:
for a start any header (H1 to H6) and for an end another header of the same type _or higher heirarchy_ or the end of the text in the node._
If you copy the text below and assign header formatting, run TOC, then try collapsing H2 A. Rather than collapse up to the next H1 - (H1 B), it will collapse up to H2 B. The resulting collapse will hide H1 B, which it should not or the hierachy structure will be confusing. This is the way that it works in Microsoft Word, which I beleive is the correct way that it should operate.

**H1 A**
Body text
Body text


**_H2 A_**
Body text
Body text
Body text


H3 A
Body text
Body text

H3 B
Body text

**H1 B**
Body text
Body text

**_H2 B_**
Body text
Body text
Body text

H3 C
Body text
Body text
Body text

**_H2 C_**
Body text
Body text

H3 D
Body text
Body text

**H1 C**
Body text

H3 E
Body text
Body text

**_H2 D_**
Body text
Body text

**_H2 E_**
Body text
Body text

**Comment by DiagonalArg** [comment]
> "...for a start any header (H1 to H6) and for an end another header of the same type or higher heirarchy or the end of the text in the node."

That's also the way it works in Org Mode (emacs).

**Comment by RFImages** [comment]
Hello again @giuspen 

1. **Collapser icon**

I have installed 1.3.0 and thank you for fixing the cursor change over the header collapser. I see a different collapsed/uncollapsed icon ( + and - ) for the collapser but because thay are the same shape, very similar, and small on a high resolution monitor I really struggle to tell the difference between collapsed and uncollapsed. I think the right arrow/down arrow icons that  you use in the Tree would be more consistent and visible. I looked in preferences/Special Characters but did not see an option to change this.

2. **alt+up and alt+ down bug**

Also the bug where moving text through a collapsed header with [alt-up arrow/down arrow] still remains. The text in the header body gets hidden. 

3. **Remove formatting**

I  feel it is very important that the "remove Formatting" also needs to remove the collapsers, since there are no longer headers..

4. **End of Body definition**

I see that the collapser will collapse up until the next header of the same type, but it should also collapse up until a header of a higher priority. i.e. if a H4 is to be collapsed, it should not collapse up until the hext H4 if there is an H3 in between, it should collapse up to the H3 instead. Otherwise collapsing it will hide the H3, which is more important to be visible.

Those are the most important ones, a few problems remain.

**Comment by giuspen** [comment]
Hello @RFImages I read carefully your points and I agree with you on the necessary improvements.
I will proceed in the next days and update here.

**Comment by RFImages** [comment]
Thank you so much @giuspen ! So happy to hear. I will make another donation when implemented. Then I will make a few more less critical requests.
I forgot to suggest putting the expander collapser icon selection in the preferences, but I personally would be very happy with the same down arrow/right arrow as used in the node tree and also Microsoft Word for this purpose. For extra credit, a space before the end of text and the expander / collapser would make it easier to target but I understand that might have complications.

**Comment by giuspen** [comment]
hello @RFImages please try https://www.giuspen.net/software/cherrytree_1.3.0.1+11_win64_portable.7z where I have addressed your points 1,2,4 while I'm thinking for point 3 to not put the clearing of the TOC anchors inside of the "remove formatting" but rather in a dedicated new menu item/action

**Comment by RFImages** [comment]
Thank you for that! However I am on Windows and not sure how to install without an .exe?

**Comment by hopliteFA** [comment]
Hi RFImages - It is a 7zip compressed file of the portable version of CT.  

To open it:

- download the 7zip Windows installer from [7zip.org](https://7-zip.org/download.html)  - you probably want the very top one on that page
- use 7zip to unzip the file by opening 7zip and then use 7zip to open the file giuspen posted
- Double click on the exe file and see if the changes work like you expected/requested.

**Comment by RFImages** [comment]
Thank you @hopliteFA , yes I have extracted the file but I find no .exe within.

**Comment by hopliteFA** [comment]
Do you see it in:

cherrytree_1.3.0.1+11_win64_portable --> mingw64m--> bin --> cherrytree.exe

**Comment by RFImages** [comment]
Yes! Thank you @hopliteFA

**Comment by RFImages** [comment]
@giuspen I have tested for items 1,2 and 4 and they appear to be working as expected. Thank you this makes it very usable and has increased my confidence. The arrow collapsers are much more visible and targetable.

Regarding item 3 , the reason I suggested removing the collapser with "Remove Formatting" is that once "Remove Formatting" eliminates all headers, the collapsers are no longer relevant and should not be still be functional, as there are no longer headers. They do not appear to function as they did prior to the formatting removal as well. 

They would be collapsing and expanding something that no longer exists (a header) and also the correct end of the collapsible section is no longer identifiable, other than historically, which could get confusing? Is there any reason they should remain? It seems that "Remove Formatting" should mean remove everything that is not plain unformatted text.

**Comment by giuspen** [comment]
Hi @RFImages 
A) you are right, I will remove the anchors in that context
B) I'll have a look if I can intercept a copy/cut and expand so that the user cannot copy unexpanded
C) makes sense, just I'm not sure about adding this in this release as that requires an update of the translations / contact all the translators, I'll have a think and see to put that with other changes that require their intervention to not bother them too much
D) thank you very much for your generosity

**Comment by RFImages** [comment]
@giuspen Prior to your next release, the more I use this the more I think it would be helpful to be able to "Expand selected area" and "Collapse selected area" in the right click menu, which would of course expand or collapse all the headers in the selected area. If this is not too difficult I would love to see it in your next release!

**Comment by giuspen** [comment]
update on B) - I changed tactic to support copy/cut of a collapsed header (extending the selection to the hidden text in that context) rather than exapanding the header.

**Comment by RFImages** [comment]
@giuspen Nice! I think that is a better solution and would make it easier to use with less distraction and disruption.

**Comment by RFImages** [comment]
Hello @giuspen well I've never heard it called "spoilers" before.
I updated to 1.4.0 and am enjoying the "spoilers" more than ever. It is nice to be able to clean everything (including the "collapsers") with Remove Formatting. 
However I have discovered a danger, and that is that if **Remove Formatting** is used on a collapsed header, the body of the Header is also removed. This another reason why I was hoping for an "**Expand/Collapse selected text**" option, it would make it a lot easier to expand multiple or all Headers with one menu click. I am finding it frustrating to expand or collapse many headers one at a time. Now if one is missed before removing formatting, there is the danger of losing text.

**Comment by giuspen** [comment]
Hi @RFImages I already implemented the right click menu items to expand/collapse selected text, I just couldn't put that int the release because that need proper translation in all languages and that will take time.
Furthermore before contacting the translators there are other changes I want to apply.
Still I can make a non official build for you to test.
Let me have a look first of this issue of the collapsed text being removed in the remove formatting as that is really unexpected since the invisibility is also a formatting tag which should be removed showing the hidden text again.

**Comment by RFImages** [comment]
@giuspen ok let me know if you agree with my finding and I am happy to test.

**Comment by giuspen** [comment]
Yes @RFImages I was able to reproduce the issue you reported in clearing the formatting and should have fixed it.
Please test https://www.giuspen.net/software/cherrytree_1.4.0.0+9_win64_portable.7z where also the right click menu items to expand/collapse the selection have been implemented.

**Comment by RFImages** [comment]
Thank you, I tested it and found that the issue is fixed. However I cannot find the expand/collapse selection in the right click menu?

**Comment by siwei-li** [comment]
Following updates on this, as I'm super stoked to have this feature, can also help with testing for MacOS (ARM) and Linux versions

**Comment by giuspen** [comment]
Hi @RFImages , it is currently a submenu of Format. Probably it is not a good choise and I should move it out.
Hi @siwei-li I'm glad you like it, I will prepare the AppImage and link it here and under https://www.giuspen.net/cherrytree/#testing

**Comment by giuspen** [comment]
( https://www.giuspen.net/cherrytree/#testing )

linuxappimage packageAppImage [CherryTree-1.4.0+9-x86_64.AppImage](https://www.giuspen.net/software/CherryTree-1.4.0+9-x86_64.AppImage) (all in one file, after downloading make it executable and then run it from a terminal or just double click)

7z archive Microsoft windows portable archive: [cherrytree_1.4.0.0+9_win64_portable.7z](https://www.giuspen.net/software/cherrytree_1.4.0.0+9_win64_portable.7z) (look for mingw64/bin/cherrytree.exe; for portable config put config.cfg in portable folder root, beside license.txt)

**Comment by RFImages** [comment]
@giuspen  Ah yes there it is. This is wonderfful  I will give it some thorough testing, meanwhile I will toss another donation your way for your trouble!

**Comment by giuspen** [comment]
Thanks @RFImages you have already been very generous, besides this is proving to be useful to other users :)

**Comment by RFImages** [comment]
I sure hope so, I am just happy for the chance to have input for how it works. I hope you find it useful as well.

**Comment by giuspen** [comment]
I'm not using the collapsible headers actively at the moment as most of my notes are actually code snippets / scripts, but I see that it is coming out nicely and I will try to make use of it to show/hide codeboxes in future :)

**Comment by RFImages** [comment]
@giuspen But where do you keep your To Do list? You must have a thousand things to do, just with the software alone. I would think you would have at least a collapsible header for "BUGS" with subheadings for each bug then sub-subheadings for "critical" "unresolved" "priority" "theory" "status" "try" etc... and "FEATURE REQUESTS" with subheadings each one, say for "priority", "status", "todo",  etc...  

There are still a few small quirks that I will send you after thoroughly testing - some odd minor things that are not at all catostrophic and easily resolved, I have not figured aout when it happens or why, so no test case. I am impressed that even images are being collapsed! I was about to report that as a problem! But overall this is completely usable and complete. Thanks so much! I hope people find it as helpful as I do. I will soon be on to annoying you about my other requests, most importantly dual panes for comparing two tabbed nodes side by side  ;)

**Comment by DiagonalArg** [comment]
Agreed on the dual panes.  I would need them for moving material from one note to another.  Sometimes, as notes are developing, we find we need to break them out into multiple smaller sub-notes.  

Or, would it be possible to just allow running two instances on the same note db?  Would that work?

**Comment by RFImages** [comment]
I think running two instances would be clumsy, as you have instances of other things that you have to work around and always be resizing the windows. I think the way Notepad++ handles it is ideal. Dual panes and each opened file (node on this case) has a tab and can be cloned or moved from one pane to the other. @giuspen  is there a thread for the dual pane idea? Don't want to get off topic.

**Comment by giuspen** [comment]
@DiagonalArg dual pane is going to happen, likely after initial support for https://github.com/giuspen/cherrytree/issues/425
@RFImages I have https://github.com/giuspen/cherrytree/issues/2614 and also https://github.com/giuspen/cherrytree/issues/2437

**Comment by RFImages** [comment]
ok here is a bug (noncatostrophic) I have noticed occasionally, and is usually fixed by collapsing/expanding the offending header again, but it became very noticable when I did a ctr-A (select all) and a collapse selection. It leaves collapsers from the sub headers within like this:

![Image](https://github.com/user-attachments/assets/a178b794-3e7e-4b78-b1a9-94d794d42e23)

If I were to now expand/collapse the header Title "Landscape" again these would go away.

**Comment by giuspen** [comment]
Hey @RFImages your tests are very useful, if you have a test document I can use to reproduce the issue, that would help. You can also send me privately via email.

**Comment by RFImages** [comment]
@giuspen, thank you for your attention to this, please be mindful that it is not only when clicking a link but also when searching for an occurrence of text, that the found text may be hidden in a collapsed header. The text highlighting is actually a separate feature request, (maybe it is or should be in another thread) but would very useful in this situation.  I find it very useful in Notepad++ especially with coding, when all occurrences are highlighted in tandem with what was searched for or manually highlighted (more likely to happen in code).

**Comment by jonathonmckay** [comment]
I am finding that now if I click a link in a TOC, it will expand any collapsed headers (if required), and select the anchor, but it doesn't scroll the note view down to the location of the anchor. (Running version 1.6.1, Windows 11)

**Comment by giuspen** [comment]
Thanks @jonathonmckay I see the regression, also reported in https://github.com/giuspen/cherrytree/issues/2768 - I'm working on it

**Comment by sadiogo** [comment]
You can manually create toggle headers by adding an anchor, naming it `h1-1`, then cutting the anchor, repasting it, writing a text, and toggling H1 property. See an example below:

<img width="1080" height="500" alt="Image" src="https://github.com/user-attachments/assets/eaa5b6bd-6057-42aa-bc51-3b315d070783" />


If you create another anchor `h1-2` below the first anchor (using the same strategy), then toggling H1-1 header will only collapse/expand up to that point.

<img width="1080" height="500" alt="Image" src="https://github.com/user-attachments/assets/acb06f98-893c-410b-9f5e-ba3b97ba8826" />
